File: Serenity.Script.UI/Editor/RadioButtonEditor.cs
using jQueryApi;
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Editor, DisplayName("Radio Button")]
    [Element("<div/>")]
    public class RadioButtonEditor : Widget<RadioButtonEditorOptions>, IStringValue
    {
        static RadioButtonEditor()
        {
            Q.Prop(typeof(RadioButtonEditor), "value");
        }

        public RadioButtonEditor(jQueryObject input, RadioButtonEditorOptions opt)
            : base(input, opt)
        {
            if (options.EnumKey.IsEmptyOrNull() &&
                options.EnumType == null &&
                options.LookupKey.IsEmptyOrNull())
                return;

            if (!options.LookupKey.IsEmptyOrNull())
            {
                var lookup = Q.GetLookup<object>(options.LookupKey);
                foreach (dynamic item in lookup.Items)
                {
                    object textValue = ((dynamic)item)[lookup.TextField];
                    var text = textValue == null ? "" : textValue.ToString();
                    object idValue = item[lookup.IdField];
                    string id = idValue == null ? "" : idValue.ToString();
                    AddRadio(id, text);
                }
            }
            else
            {
                var enumType = options.EnumType ?? EnumTypeRegistry.Get(options.EnumKey);

                var enumKey = options.EnumKey;
                if (enumKey == null && enumType != null)
                {
                    var enumKeyAttr = enumType.GetCustomAttributes(typeof(EnumKeyAttribute), false);
                    if (enumKeyAttr.Length > 0)
                        enumKey = enumKeyAttr[0].As<EnumKeyAttribute>().Value;
                }

                foreach (var x in Enum.GetValues(enumType))
                {
                    var name = Enum.ToString(enumType, x.As<Enum>());
                    AddRadio(((int)x).ToString(), Q.TryGetText("Enums." + enumKey + "." + name) ?? name);
                }
            }
        }

        private void AddRadio(string value, string text)
        {
            var label = J("<label/>").Text(text);
            J("<input type=\"radio\"/>")
                .Attribute("name", this.uniqueName)
                .Attribute("id", this.uniqueName + "_" + value)
                .Attribute("value", value)
                .PrependTo(label);
            label.AppendTo(this.element);
        }

        public string Value
        {
            get
            {
                return this.element.Find("input:checked").First().GetValue();
            }
            set
            {
                if (value != this.Value)
                {
                    var inputs = this.element.Find("input");
                    inputs.Filter(":checked").RemoveAttr("checked");
                    if (!string.IsNullOrEmpty(value))
                        inputs.Filter("[value=" + value + "]").Attribute("checked", "checked");
                }
            }
        }
    }

    [Serializable, Reflectable]
    public class RadioButtonEditorOptions
    {
        public RadioButtonEditorOptions()
        {
        }

        public String EnumKey { get; set; }
        public Type EnumType { get; set; }
        public String LookupKey { get; set; }
    }
}

File: Serenity.Scripts/CoreLib/UI/Editors/RadioButtonEditor.ts
namespace Serenity {

    export interface RadioButtonEditorOptions {
        enumKey?: string;
        enumType?: any;
        lookupKey?: string;
    }

    @Decorators.registerEditor('Serenity.RadioButtonEditor', [IStringValue, IReadOnly])
    @Decorators.element('<div/>')
    export class RadioButtonEditor extends Widget<RadioButtonEditorOptions>
        implements IReadOnly {

        constructor(input: JQuery, opt: RadioButtonEditorOptions) {
            super(input, opt);

            if (Q.isEmptyOrNull(this.options.enumKey) &&
                this.options.enumType == null &&
                Q.isEmptyOrNull(this.options.lookupKey)) {
                return;
            }

            if (!Q.isEmptyOrNull(this.options.lookupKey)) {
                var lookup = Q.getLookup(this.options.lookupKey);
                for (var item of lookup.items) {
                    var textValue = item[lookup.textField];
                    var text = (textValue == null ? '' : textValue.toString());
                    var idValue = item[lookup.idField];
                    var id = (idValue == null ? '' : idValue.toString());
                    this.addRadio(id, text);
                }
            }
            else {
                var enumType = this.options.enumType || Serenity.EnumTypeRegistry.get(this.options.enumKey);
                var enumKey = this.options.enumKey;
                if (enumKey == null && enumType != null) {
                    var enumKeyAttr = Q.getAttributes(enumType, Serenity.EnumKeyAttribute, false);
                    if (enumKeyAttr.length > 0) {
                        enumKey = enumKeyAttr[0].value;
                    }
                }

                var values = Q.Enum.getValues(enumType);
                for (var x of values) {
                    var name = Q.Enum.toString(enumType, x);
                    this.addRadio(x.toString(), Q.coalesce(Q.tryGetText(
                        'Enums.' + enumKey + '.' + name), name));
                }
            }
        }

        protected addRadio(value: string, text: string) {
            var label = $('<label/>').text(text);
            $('<input type="radio"/>').attr('name', this.uniqueName)
                .attr('id', this.uniqueName + '_' + value)
                .attr('value', value).prependTo(label);
            label.appendTo(this.element);
        }

        get_value(): string {
            return this.element.find('input:checked').first().val();
        }

        get value(): string {
            return this.get_value();
        }

        set_value(value: string): void {
            if (value !== this.get_value()) {
                var inputs = this.element.find('input');
                var checks = inputs.filter(':checked');
                if (checks.length > 0) {
                    (checks[0] as HTMLInputElement).checked = false;
                }
                if (!Q.isEmptyOrNull(value)) {
                    checks = inputs.filter('[value=' + value + ']');
                    if (checks.length > 0) {
                        (checks[0] as HTMLInputElement).checked = true;
                    }
                }
            }
        }

        set value(v: string) {
            this.set_value(v);
        }

        get_readOnly(): boolean {
            return this.element.attr('disabled') != null;
        }

        set_readOnly(value: boolean): void {
            if (this.get_readOnly() !== value) {
                if (value) {
                    this.element.attr('disabled', 'disabled')
                        .find('input').attr('disabled', 'disabled');
                }
                else {
                    this.element.removeAttr('disabled')
                        .find('input').removeAttr('disabled');
                }
            }
        }

    }
}